File: homing-workspace-shell/src/main/java/hue/captains/singapura/js/homing/workspace/shell/CodecRegistrarModule.java
package hue.captains.singapura.js.homing.workspace.shell;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;
import hue.captains.singapura.js.homing.core.ModuleImports;
import hue.captains.singapura.js.homing.workspace.persistence.WidgetParamsCodecRegistryModule;

import java.util.List;

/**
 * {@code CodecRegistrar} — Phase 3 of the workspace-shell chrome.
 * Two-step registration of widget Params codecs:
 *
 * <ol>
 *   <li><b>Defaults (sync):</b> every spec {@code entry} gets the
 *       substrate's identity codec ({@code transformTo}/{@code transformFrom}
 *       both return {@code {}}). Covers widgets whose Params are
 *       {@code Widget._None} or otherwise round-trip-trivial — the
 *       majority.</li>
 *   <li><b>Custom (async overlay):</b> each entry in
 *       {@code spec.widgetCodecs} ({@link WidgetCodecRef}) is
 *       dynamic-imported and registered under its
 *       {@code widgetKind}, overriding the identity codec for that
 *       kind. Tested in isolation: the synchronous
 *       {@code _registerDefaults} + {@code _registerCustom} cores are
 *       directly callable.</li>
 * </ol>
 *
 * <p>Replaces the inline registration loop V1 carries at chrome lines
 * 244–297. Must run BEFORE persistence attach (Phase 4) — saved-state
 * round-trip dispatches by widget kind through this registry, so an
 * unregistered kind marks restore failed.</p>
 *
 * <p>Explicit Substrate doctrine: instance methods on INSTANCE
 * singleton; no static methods; {@code registerAll} returns
 * {@code Promise<void>} explicitly (the I/O point is the dynamic
 * imports inside).</p>
 *
 * @since post-RFC-0034 workspace chrome decomposition — Phase 3
 */
public record CodecRegistrarModule() implements DomModule<CodecRegistrarModule> {

    public static final CodecRegistrarModule INSTANCE = new CodecRegistrarModule();

    /** The {@code CodecRegistrar} JS class. */
    public record CodecRegistrar() implements Exportable._Class<CodecRegistrarModule> {}

    @Override
    public ImportsFor<CodecRegistrarModule> imports() {
        return ImportsFor.<CodecRegistrarModule>builder()
                .add(new ModuleImports<>(
                        List.of(new WidgetParamsCodecRegistryModule.WidgetParamsCodecRegistry()),
                        WidgetParamsCodecRegistryModule.INSTANCE))
                .build();
    }

    @Override
    public ExportsOf<CodecRegistrarModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new CodecRegistrar()));
    }
}

File: homing-workspace-shell/src/main/resources/homing/js/hue/captains/singapura/js/homing/workspace/shell/CodecRegistrarModule.js
// =============================================================================
// CodecRegistrar — Phase 3 of the workspace-shell chrome.
//
// Two-step registration of widget Params codecs:
//   1. Defaults (sync): every spec.entries[i] gets the substrate's
//      identity codec.
//   2. Custom (async overlay): spec.widgetCodecs entries are dynamic-
//      imported and registered under widgetKind.
//
// Explicit Substrate doctrine:
//   - Instance methods on INSTANCE singleton (no static).
//   - No async sugar — Promises returned explicitly.
//   - ALL collaborators in constructor: _registry, _importer.
//
// Instance fields (graph-traversable):
//   _registry : WidgetParamsCodecRegistry — defaults to the global
//   _importer : (url) => Promise<module>  — defaults to native dynamic import
// =============================================================================

class CodecRegistrar {

    constructor(deps) {
        deps = deps || {};
        this._registry = deps.registry || WidgetParamsCodecRegistry;
        this._importer = deps.importer || (url => import(url));
    }

    /**
     * Two-step registration. Synchronous defaults fire immediately;
     * custom overlays fire when their dynamic imports resolve.
     *
     * @return Promise<void>
     */
    registerAll(opts) {
        if (!opts) throw new Error('[CodecRegistrar] opts required');

        // 1. Sync defaults.
        this._registerDefaults(opts.entries || []);

        // 2. Async custom overlay.
        const refs = opts.widgetCodecs || [];
        if (refs.length === 0) return Promise.resolve();
        const self = this;
        return Promise.all(refs.map(r => self._importer(r.moduleUrl)))
            .then(modules => self._registerCustom(refs, modules));
    }

    /** SYNC: register identity codec for each entry's simpleName. */
    _registerDefaults(entries) {
        const identity = this._identityCodec();
        for (const e of entries) {
            this._registry.register(e.simpleName, identity);
        }
    }

    /** SYNC: walk (refs, modules) pairwise, register each codec under
     *  its widgetKind. Skips refs whose module is missing the export. */
    _registerCustom(refs, modules) {
        for (let i = 0; i < refs.length; i++) {
            const mod = modules[i];
            const ref = refs[i];
            const codec = mod && mod[ref.exportName];
            if (!codec) {
                console.error('[CodecRegistrar] codec export "' + ref.exportName
                            + '" not found in module ' + ref.moduleUrl
                            + ' — widget kind "' + ref.widgetKind
                            + '" stays on identity codec.');
                continue;
            }
            if (typeof codec.transformTo !== 'function'
             || typeof codec.transformFrom !== 'function') {
                console.error('[CodecRegistrar] codec for "' + ref.widgetKind
                            + '" missing transformTo / transformFrom — skipped.');
                continue;
            }
            this._registry.register(ref.widgetKind, codec);
        }
    }

    /** The substrate's no-op Params codec: empty in, empty out. */
    _identityCodec() {
        return {
            transformTo:   function (_p) { return {}; },
            transformFrom: function (_w) { return {}; }
        };
    }
}

CodecRegistrar.INSTANCE = new CodecRegistrar();
